chromadb/api/models/CollectionCommon.py

```diff
@@ -461,41 +428,64 @@ def _validate_and_prepare_update_request(
         ],
         metadatas: Optional[OneOrMany[Metadata]],
         documents: Optional[OneOrMany[Document]],
-        images: Optional[OneOrMany[Image]],
-        uris: Optional[OneOrMany[URI]],
-    ) -> Tuple[
-        IDs,
-        Embeddings,
-        Optional[Metadatas],
-        Optional[Documents],
-        Optional[URIs],
-    ]:
-        (
-            ids,
-            embeddings,
-            metadatas,
-            documents,
-            images,
-            uris,
-        ) = self._validate_embedding_set(
-            ids,
-            embeddings,
-            metadatas,
-            documents,
-            images,
-            uris,
-            require_embeddings_or_data=False,
+        images: Optional[OneOrMany[Image]] = None,
+        uris: Optional[OneOrMany[URI]] = None,
+    ) -> RecordSet:
+        unpacked_embedding_set = self._unpack_record_set(
+            ids=ids,
+            embeddings=embeddings,
+            metadatas=metadatas,
+            documents=documents,
+            images=images,
+            uris=uris,
         )
 
+        normalized_embeddings = (
+            self._normalize_embeddings(unpacked_embedding_set["embeddings"])
+            if unpacked_embedding_set["embeddings"] is not None
+            else None
+        )
+
+        self._validate_embedding_set(
+            ids=unpacked_embedding_set["ids"],
+            embeddings=normalized_embeddings,
+            metadatas=unpacked_embedding_set["metadatas"],
+            documents=unpacked_embedding_set["documents"],
+            images=unpacked_embedding_set["images"],
+            uris=unpacked_embedding_set["uris"],
+        )
+
+        prepared_embeddings = self._compute_embeddings(
+            embeddings=normalized_embeddings,
+            documents=unpacked_embedding_set["documents"],
+            images=unpacked_embedding_set["images"],
+            uris=unpacked_embedding_set["uris"],
+        )
+
+        return {
+            "ids": unpacked_embedding_set["ids"],
+            "embeddings": prepared_embeddings,
+            "metadatas": unpacked_embedding_set["metadatas"],
+            "documents": unpacked_embedding_set["documents"],
+            "images": unpacked_embedding_set["images"],
+            "uris": unpacked_embedding_set["uris"],
+        }
+
+    def _compute_embeddings_upsert_or_update_request(
```

Comment:
the only thing which is different between this and `_compute_embeddings` is that this does not take `uris` as an argument. Since we validate the presence of necessary fields before this step, I think we can get rid of this redundancy, and instead re-use `_compute_embeddings`, passing `None` for the `uris` field.

In the case where we pass all `None` the error string would be `You must provide either embeddings, documents, images, or uris.` which might be slightly confusing since update / upsert does not take `uris`, but is more like an internal error for us than something the user did wrong.

Reply:
Taking a look through the various code-paths; the reason we (typically) want to avoid updating a URI is because it might point to a different data-type than the original. This needs more thought, let's open an issue, but also let's make this change.

Reply:
https://github.com/chroma-core/chroma/issues/2744